graphql_server/render_graphiql.py

```diff
@@ -0,0 +1,306 @@
+    """
+
+    graphiql_version: Optional[str]
+    graphiql_template: Optional[str]
+    graphiql_html_title: Optional[str]
+    jinja_env: Optional[Environment]
+
+
+class GraphiQLOptions(TypedDict):
+    """GraphiQL options to display on the UI.
+
+    Has the following attributes:
+
+    default_query
+        An optional GraphQL string to use when no query is provided and no stored
+        query exists from a previous session.  If undefined is provided, GraphiQL
+        will use its own default query.
+    header_editor_enabled
+        An optional boolean which enables the header editor when true.
+        Defaults to false.
+    """
+
+    default_query: Optional[str]
+    header_editor_enabled: Optional[bool]
+
+
+def escape_js_value(value: Any) -> Any:
+    quotation = False
+    if value.startswith('"') and value.endswith('"'):
+        quotation = True
+        value = value[1 : len(value) - 1]
+
+    value = value.replace("\\\\n", "\\\\\\n").replace("\\n", "\\\\n")
+    if quotation:
+        value = '"' + value.replace('\\\\"', '"').replace('"', '\\"') + '"'
+
+    return value
+
+
+def process_var(template: str, name: str, value: Any, jsonify=False) -> str:
+    pattern = r"{{\s*" + name + r"(\s*|[^}]+)*\s*}}"
+    if jsonify and value not in ["null", "undefined"]:
+        value = json.dumps(value)
+        value = escape_js_value(value)
+
+    return re.sub(pattern, value, template)
+
+
+def simple_renderer(template: str, **values: Dict[str, Any]) -> str:
+    replace = [
+        "graphiql_version",
+        "subscription_url",
+        "header_editor_enabled",
+        "graphiql_html_title",
+    ]
+    replace_jsonify = [
+        "query",
+        "result",
+        "variables",
+        "operation_name",
+        "default_query",
+    ]
+
+    for r in replace:
+        template = process_var(template, r, values.get(r, ""))
+
+    for r in replace_jsonify:
+        template = process_var(template, r, values.get(r, ""), True)
+
+    return template
+
+
+def _render_graphiql(
+    data: GraphiQLData,
+    config: Optional[GraphiQLConfig] = None,
+    options: Optional[GraphiQLOptions] = None,
+) -> Tuple[str, Dict[str, Any]]:
+    """When render_graphiql receives a request which does not Accept JSON, but does
+    Accept HTML, it may present GraphiQL, the in-browser GraphQL explorer IDE.
+    When shown, it will be pre-populated with the result of having executed
+    the requested query.
+    """
+    if config is None:
+        config = {}
+    graphiql_version = config.get("graphiql_version") or GRAPHIQL_VERSION
+    graphiql_template = config.get("graphiql_template") or GRAPHIQL_TEMPLATE
+    graphiql_html_title = config.get("graphiql_html_title") or "GraphiQL"
+
+    template_vars = {
+        "graphiql_version": graphiql_version,
+        "graphiql_html_title": graphiql_html_title,
+        "query": data.get("query"),
+        "variables": data.get("variables"),
```

Comment:
we also have a new `headers` string prop that allows you to set the default `headers` value if you want :)

Reply:
I just noticed latest changes on [renderGraphiQL.js](https://github.com/graphql/express-graphql/blob/master/src/renderGraphiQL.js) have the `headers` option but it is passed at the `graphQLFetcher` function whereas these variables are passed to the ReactDom function.

Reply:
you can see the new props here: https://github.com/graphql/graphiql/tree/1.0.0/packages/graphiql#props

the headers that are passed in the fetcher function are the ones from the editor. the `headers` react prop is for pre-populating the headers editor